### PR TITLE
#4501

### LINKED ISSUE #4431: Smelter incorrectly shows as "Smelters" in the Town Hall
### Minecolonies version
Version: 
minecolonies-0.10.761-ALPHA-universal.jar
structurize-0.10.279-RELEASE.jar
Minecraft Version: 1.15.2 (modded) forge-31.1.19

### Expected behavior
- Town Hall Information GUI shows: 
"[...]
Smelter 1/1
[...]"

### Actual behaviour
- Town Hall Information GUI shows:
"[...]
Smelter 0/1
Smelters 1/null
[...]"

It seems as if the Smeltery worker is registered as "Smelters" while the Smeltery itselfs registers as needing a "Smelter"

### Steps to reproduce the problem

1. Build Town Hall and Smeltery with worker
2. Check Town Hall GUI

PS. It also shows "Children 1/null which probably isn't a problem, just looks kind of weird

### LINKED ISSUE #4490: Crafter Death or Firing
### Minecolonies version

Version: minecolonies-0.11.785-BETA-universal

### Expected behavior

- When a Crafter dies or otherwise leaves the building, all current requests to that building are cancelled and new ones are not made.

### Actual behaviour

- When a Crafter (seen with mason, miller, and blacksmith) is killed by raiders, the Builder request for their work remains. The Builder will sit and wait for the parts that will never come, without the "needs attention" exclamation-gear. Also, I think it is related, workers keep requesting tools from the Blacksmith's Shop even if the Blacksmith has been fired or killed. 

### Steps to reproduce the problem

1. Have a Builder request from a Crafter
2. Kill the Crafter
3. Builder request does not cancel
4. ... Profit

1. Hire Blacksmith
2. Fire or kill Blacksmith
3. Miner will still ask Blacksmith to craft iron pickaxes
4. ... Profit

### Logs: Latest.log and crashlog at https://gist.github.com/

- latest.log:
- crashlog:

### LINKED ISSUE #4500: Feature Request: Reset request system at town hall
### Minecolonies version

Version: minecolonies-0.11.785-BETA-universal

### Feature request

- Resetting the request system is still sometimes required. Using the /mc colony command to reset the request system requires cheats enabled. For those of us who really want to commit to Survival without cheats, can you add a button to the Town Hall block that resets the request system without cheats having to be enabled? 

Thanks.

### LINKED ISSUE #4501: Citizens dont drop items when killed
### Minecolonies version

Version: 
minecolonies-0.11.788-ALPHA-universal.jar
forge 1.15.2 - 31.1.27

### Expected behavior

When a citizen gets killed (unfortunely) it drops all the items inside his inventory.

### Actual behaviour

The citizens, when killed by any hostile mob, dont drop anything at all.

### Steps to reproduce the problem

1.  A hostile mob targets a citizen
2. It kills him with rage
3. The citizen goes to the otherworld with all his items (pharaons will be proud)
4. ... Profit

### PR DESCRIPTION
**Pull request by Raycoms**
Closes #4501
Closes #4490
Closes #4500
Closes #4431

# Changes proposed in this pull request:
- Drop inv
- Detect firing/death to also cancel queue.
- Add request system reset button to townhall
- Fix smeltery job request in townhall list

Review please

### CODE REVIEW

**Review by someaddons — DISMISSED** [review]

**Comment by reggaemuffin** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/main/java/com/minecolonies/coremod/network/messages/ResetRequestSystemMessage.java  3
         

Clones added
============
- src/main/java/com/minecolonies/coremod/network/messages/ResetRequestSystemMessage.java  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Minecolonies/minecolonies/pullRequest?prid=5243656&bid=17187111)

**Comment by reggaemuffin** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/main/java/com/minecolonies/coremod/network/messages/ResetRequestSystemMessage.java  3
         

Clones added
============
- src/main/java/com/minecolonies/coremod/network/messages/ResetRequestSystemMessage.java  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Minecolonies/minecolonies/pullRequest?prid=5243656&bid=17187111)

**Review by marchermans — CHANGES_REQUESTED** [review]
I personally think adding a button is a step to much when it comes to resetting the RS.
The reason why it is a command, is more or less to nudge people to report issues. With this button there is no reason to do the reporting anymore.

As such I would like it to be removed again, i understand the users request to want to play pure without cheats, but this is not cheating he or she is free to turn the cheats back off after the reset.

If it happens to often for him, that he thinks it is an inconvenience he or she should report it, together with this world etc. And let us find a solution for the problem.

As such i reject the change made for issue #4500 and propose it to be reverted.

**Comment by kd5crs** [comment]
OrionDevelopment, I submitted the request for the button. I 100% want to help make the request system be all it can be, it is my favorite part of the mod. The reason I would like to not require cheats is that I play Minecraft with the kids, and they sometimes... let's say they don't always want to go find that last flower or all those diamonds. So I really want to be able to play Minecolonies with even the ability to enable cheats disabled, sort of like a hardcore mode for the mod. Personally, having it as a button won't affect my reporting issues at all, it will just make it possible to play 100% without cheats.

Here's the other reason: I don't always know if there is a RS issue or if it is something else. For example, one time I simply could not get the builder to build. I was sure it was the RS, but every time I reset it the Builder went back to the same position and stopped. It turned out that there was a hidden Pillager, and once I killed the Pillager it all started working. So even though the issue wasn't the RS, being able to reset the RS helped me find the problem. I am probably a bit reset-happy, it is true. 

Would it be possible to put the Town Hall RS reset button on a new page of the Town Hall GUI, with a message saying something like: "You should never have to use this! If you do, please open an issue with the circumstances on our github page." Maybe even write a special RS log every time it is reset?

Thanks all.

**Comment by Raycoms** [comment]
How about a compromise, @OrionDevelopment I only display the button if the "developer options" config value is true?

**Comment by marchermans** [comment]
TLDR: No the button will not be added, nobody has presented me, neither @kd5crs nor @Raycoms with a valid reason for why it should be simpler and easier to reset the rs, cause nobody can give me an argument why it is more of a burden on them to reset the rs, which they do not often, since they do not report the issue, then it is a burden to create an issue and get the problem looked at.

@kd5crs
I though this over. Initially i could see your point. But the more I thought this through the more I became against this. 
Maybe I should explain the situation better: We initially had a system that could not be reset and as such if you had a citizen that was stuck you needed to kill him or her to get it to work again. Now with the introduction of the RS the memory of what a citizen was doing (In terms of what items he needed for his task, and later on what he was crafting etc) was moved from the citizen to the RS itself. With the initial RS there was no way at all to reset it. If it was stuck there was only one way to get it fixed: Kill them. This meant there was a high potential for people to report stuff. As with all new systems, a lot was wrong, and a lot was broken. Partly due to the fact that it was my first real interaction with minecolonies and partly due to the fact that i did not yet entirely understood its inner workers and what side effects people exploited to make things work just the way they liked them, which i now broke. 

With the flood of complains that came in i added the command as a last resort. Because, well it was easy. The RS knew everything. So resetting it made for a simple fix to citizens getting stuck due to some interaction between either two citizens, a citizen and the world or just the RS it self being derpy. The goal of this command was there to give me breathing room, so that i could fix the RS. Overtime i did. But added the command also had a massive side effect. My source of information on what needed to be fixed stopped. Everytime somebody had an issue, they joined our discord, asked around and got told to run the command. So from now on this is what people did, run the command. It took me months to teach people to not suggest this command to people willy nilly. And i am extremely disgusted at my self for ever adding it. The mod is honestly worst of with it existing.

Now to your case: Your argument that you want to use the button instead of the commands to reset the RS, because your children would abuse the power of the command is flawed from the get go. It assumes that your children know when to run the reset and when to report an issue. Which they obviously do not, since 1) you do not even know that as stated and 2) they can not even restrain them self from giving themselfs diamonds if the commands are turned on using a cheat command. However this is completely understandable they are children after all. The simple fact that you however do not want to turn on cheats, then run the command, and then leave the world and join it back to turn the commands back of, suggest that you have no intention what so ever to report an issue. Since all of this: the turning on, running the command etc, takes less then 2 minutes of your time. And that is even to much for you. If this is such a burden on you, why did you not report the problems? I could not find any concrete issue you reported here on github in this direction (please post links if i missed them, i will personally try to help you with them). That said no, even the compromise @Raycoms offered is to much. That command is there for people to be a burden, especially in single player. On server there are OPs that know when to report and when to run the command. But in singleplayer i want as much info as i can get. If people think that it is too much to report an issue they find, so we can fix it in our free time, then maybe the command should be removed entirely, or kill of all citizens in the process of using it. Anyways.... I am rambling.

Hopefully, you can send me some links to issues you had, and i am able to help you with them.
If not i am sorry but that button will not be added to the UI.

**Comment by kd5crs** [comment]
I think you are fundamentally misunderstanding what I am saying starting with "Now to your case". I don't know why you jumped to being insulting. This is just a convenience button for me to allow playing Minecraft the way I like. I did not intend to tap into these underlying issues with the effect of the RS Reset command and bug reporting. It is neither my intention nor my desire to get involved with how y'all develop.

But if you are asking: please give the player more information on what's going on with the RS. Often things just stop happening, and there's no way to tell why. I figured out one issue that I put in #4490, but sometimes things just... stop. It is very common for the DMan to run between huts without actually moving anything, for example. Without any information about what the DMan thinks they are doing, there's nothing to write an Issue about. Is there an RS log file somewhere I am missing?

Thanks.

**Comment by marchermans** [comment]
@kd5crs Sorry if you feel insulted. This was not meant as an insult. Far from it. I am happy that you came to use with concerns and are actually discussing this in a constructive manner. You are one of the first to actually bring arguments to the table, why they want it to be easier.

That said, i would like to take a minute to discuss my point a bit clearer now that i slept a night on it. If I was in your position i would love this button as well. To be fair, i would love nothing more to make it so that people can play the way they want to. But indulge me for a second when I go down this rabbit hole. The following things would happen when the button is added:
1) People who normally have problems, stop reporting them. This is due to the nature of how Entities in Minecraft are more or less asynchronous. A small nudge to the parameters of the system might make it behave differently, even though the underlying problem was not solved. This is bad, very bad even. Cause it means that we can not fix the issues even though we should.
2) People who use the button regularly get frustrated when it does not solve the issue. Do not get me wrong, there will be enough people who keep testing and figure their issue out eventually. But from my experience in our discord a lot of them will also simply throw a tantrum. Pings admins during the night etc. Simply said it will not improve the culture of the community one bit. We all have to work together to get everything solved.

This all said, I still think a simple button is not what we need to solve all the issues. Yes it would solve your particular case, extremely neatly. And I might consider writing a mini mod for you that will in the end handle the case you are looking for..... But this is not a guarantee, i do not have infinite time. Additionally if I add such a mini mod, I will put a high cost on it. Since it is not meant to be used as a debug tool, simply to discourage people from blaming the RS and not helping constructively solve the issue.

Now how I think we can solve this issue better.
I would propose an extension to the Interaction system. We would need to add more information to it, that explains why the citizen is standing still. Or not picking up stuff, in case of the DMan. However we will never be able to solve all the weird quirks that come with Minecraft and its Entities. And this is something you have to understand as well.

**Comment by someaddons** [comment]
I think orion is right on this, we already have the command to hard reset it if needed(and do so on a restart aswell I think). Normally cancelling the request should be just fine to make it recalculate it when stuck.  And if it often runs into issues we do want to know, so we can fix it.

**Review by marchermans — APPROVED** [review]
